images/haproxy/push-cross.sh

```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+
+# Copyright 2019 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# builds and pushes the haproxy image for all architectures
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# cd to the repo root
+REPO_ROOT=$(git rev-parse --show-toplevel)
+cd "${REPO_ROOT}"
+
+# cd to the image
+cd ./images/haproxy
+
+IMAGE="${IMAGE:-kindest/haproxy}"
+TAG="${TAG:-2.0.0-alpine}"
+BASE="haproxy:${TAG}"
+
+# tag arch, manifest architecture, variant
+ARCHES=(
+  "amd64,amd64,"
+  "arm32v6,arm,v6"
+  "arm64v8,arm64,v8"
+  "ppc64le,ppc64le,"
+)
+
+# build all images
+images=()
+image_infos=()
+for arch_info in "${ARCHES[@]}"; do
+  # build image
+  tag_arch="$(sed -E 's#^([^,]+),[^,]*,[^,]*$#\1#' <<<"${arch_info}")"
+  image="${IMAGE}:${tag_arch}-${TAG}"
+  docker build "--build-arg=ARCH=${tag_arch}" "--build-arg=BASE=${BASE}" -f Dockerfile -t "${image}" .
+  docker push "${image}"
+  # join image we tagged with arch info for building the manifest later 
+  images+=("${image}")
+  image_infos+=("${image},${arch_info}")
+done
+
+# This option is required for running the docker manifest command
+export DOCKER_CLI_EXPERIMENTAL="enabled"
+
+# create and push the manifest
```

Comment:
why not building and creating the manifest?
just curious, I never did this before, but seems too complicated having to use all this `regex`

Reply:
Manifests are a meta image pointing to images for different hosts in this context, I'm not sure what you mean by "why not building and creating the manifest?"

Reply:
The regex is actually pretty simple and just to extract parts of the comma-joined tuple.
Each image has some metadata we need. We need to know:
- what we built & tagged the specific image as
- what architecture it was for
- what variant, if any it was for
- when building: what to pull it as (which isn't always the same as arch + variant ...)
When we build the "meta image" (manifest list image) we have to supply most of that info for each specific image we built.

Reply:
Kubernetes has some extremely similar logic for all of the cross platform images.

At some point I'll DRY this out to be nicer with kindnetd but for now this should be suitable and we only need to build if / when we want to upgrade haproxy anyhow.

Reply:
yeah, that makes sense.
For "why not building and creating the manifest? I meant to do everything in the same `for` loop, seems that the variables you are extracting are available at build time

Reply:
Oh! Because if an image build fails we leak the manifest. It only removed the manifest with the -p/--purge on pushing the manifest. We don't want to increase the chance of having a lingering manifest. You have to literally find and remove the files on disk in what I think is an undocumented directory.

Reply:
So we only build the manifest after all the images are good to go.

Reply:
Got it , thanks